Added instanceof WP_Post check

darssen opened the pull request:
Fixes #16835
>In some scenarios, the following notice can be triggered:
>>PHP Notice: Trying to get property 'post_author' of non-object in wp-content/plugins/jetpack/class.jetpack-post-images.php on line 528

#### Changes proposed in this Pull Request:
<!--- Explain what functional changes your PR includes -->
Added instanceof WP_Post check in `from_gravatar` function to avoid accessing post_author of a non-object. If that is the case, the function will return an empty array similar to what other functions in the file do (`from_blocks`, `from_images`, `from_blavatar`...)


#### Other information:

- [ X ] Have you written new tests for your changes, if applicable?
- [ X ] Have you checked the E2E test CI results, and verified that your changes do not break them?

#### Jetpack product discussion
No

#### Does this pull request change what data or activity we track or use?
No

#### Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->
Haven't been able to reproduce the issue with the normal flow of the Application so needed to do some **tweaking**. The easiest is through Related Posts. Top posts widget also seemed to go through the flow, but I have been unable to test it that way.
* Enable related posts in Jetpack 
![image](https://user-images.githubusercontent.com/13239096/186658088-becb8b30-95ef-4378-a642-0f54eac32f61.png)
* Create two posts without images attached or galleries and same tags so that they will appear as related. (It may take a while to process the related posts) 
* Go to the first post created and check that related posts are available. One of them should be the second post, and it should not show any images, it should show the text instead.
* In the `get_images` method, set to **true** `fallback_to_avatars` in the `defaults` array (line 664). This way at the end of the method, the from_gravatar method will be called if no images are associated with the related post.
* Set post to null by switching `$post = get_post( $post_id );` with `$post = null;` at the beginning of `from_gravatar` method.
With these instructions and the code added in the commit, you should not see the Notice or Warning (depending on PHP Version running). If code from the commit is commented out, then the Notice or Warning will be available.

darssen committed:
Added instanceof check to avoid accessing post_author of non WP_Post instances.

jeherve commented on projects/plugins/jetpack/changelog/class.jetpack-post-images.php:
```diff
@@ -0,0 +1,4 @@
+Significance: patch
+Type: bugfix
+
+Post Images: Ensure $post is a WP_Post instance.
```
I would recommend being a bit more generic in here, since that changelog entry will be displayed to all end-users when they update the Jetpack plugin. The entry consequently has to be as user-friendly as possible. How about something like:

```suggestion
Post Images: avoid PHP notices when fetching images from posts with missing metadata.
```

darssen replied:
Yes, this makes much more sense. Thanks for the suggestion!

jeherve commented on projects/plugins/jetpack/changelog/class.jetpack-post-images.php:
```diff
@@ -0,0 +1,4 @@
+Significance: patch
```
Do you happen to remember how you created that file? If you run `jetpack changelog add plugins/jetpack`, the changelog file name usually uses the branch name by default. You wouldn't need a php file extension here since that's just text.

darssen replied:
>Do you happen to remember how you created that file?

Yeah, it automatically prompted me for the changelog while pushing my branch upstream. I was confused by the 
`Name your change file `So I switched to the name of the file :)

Checking the rest of the changelogs I see the expected naming.

jeherve replied:
> Name your change file So I switched to the name of the file :)

That's good feedback. Now that you've gone through the process, would you phrase that prompt differently?

darssen replied:
I guess for me it would have been more intuitive to have `Name your changelog file`

jeherve reviewed (commented):
This make sense. I realize we do not have any tests for the `from_gravatar` method in our tests (see `projects/plugins/jetpack/tests/php/media/test-class.jetpack-post-images.php`) right now. Maybe this would be the opportunity to add a test that would cover this scenario (messed up $post) as well as the expected successful return? What do you think?

darssen committed:
Improved changelog description

Co-authored-by: Jeremy Herve <jeremy@jeremy.hu>

darssen committed:
Updated changelog file name to follow guidelines

darssen committed:
Added unit tests to for_gravatar function

jeherve commented on projects/plugins/jetpack/tests/php/media/test-class.jetpack-post-images.php:
```diff
@@ -633,4 +633,53 @@ public function test_from_story_block_from_post_id_is_correct_array_videopress_w
+	 * Test if the array extracted is empty in case post_id is null.
+	 *
+	 * @covers Jetpack_PostImages::from_gravatar
+	 * @since 11.3
```
Jetpack includes tooling that allows you to make this dymamic. This way you do not have to know or think about when your change will be merged, you can let the tooling update that number for you in time with the next release.
https://github.com/Automattic/jetpack/blob/3a3c4adc1c0913bac224ddb832713bd6101fe1ab/projects/packages/README.md#L109-L116

```suggestion
	 * @since $$next-version$$
```

darssen replied:
Oh, that's good to know :)

jeherve commented on projects/plugins/jetpack/tests/php/media/test-class.jetpack-post-images.php:
```diff
@@ -633,4 +633,53 @@ public function test_from_story_block_from_post_id_is_correct_array_videopress_w
+	 * Test if the array extracted is empty in case post_id is an integer.
+	 *
+	 * @covers Jetpack_PostImages::from_gravatar
+	 * @since 11.3
```
```suggestion
	 * @since $$next-version$$
```

jeherve commented on projects/plugins/jetpack/tests/php/media/test-class.jetpack-post-images.php:
```diff
@@ -633,4 +633,53 @@ public function test_from_story_block_from_post_id_is_correct_array_videopress_w
+	 * Test if the array extracted has a valid image when sending a valid post.
+	 *
+	 * @covers Jetpack_PostImages::from_gravatar
+	 * @since 11.3
```
```suggestion
	 * @since $$next-version$$
```

jeherve commented on projects/plugins/jetpack/tests/php/media/test-class.jetpack-post-images.php:
```diff
@@ -633,4 +633,53 @@ public function test_from_story_block_from_post_id_is_correct_array_videopress_w
 		$this->assertEquals( $expected_poster_url, $images[1]['src'] );
 	}
+
+	/**
```
Thanks for adding tests! ❤️ 

This test file has been around for quite some time, and is not very optimized. As you're adding a new test, I think it would make sense to switch to using a data provider method so you do not have to repeat things too much. Here is how you could do it:

```diff
diff --git a/projects/plugins/jetpack/tests/php/media/test-class.jetpack-post-images.php b/projects/plugins/jetpack/tests/php/media/test-class.jetpack-post-images.php
index f85335b417..eeea9e841e 100644
--- a/projects/plugins/jetpack/tests/php/media/test-class.jetpack-post-images.php
+++ b/projects/plugins/jetpack/tests/php/media/test-class.jetpack-post-images.php
@@ -635,51 +635,52 @@ class WP_Test_Jetpack_PostImages extends WP_UnitTestCase {
 	}
 
 	/**
-	 * Test if the array extracted is empty in case post_id is null.
+	 * Test if the array extracted is empty in case post_id is a string.
 	 *
 	 * @covers Jetpack_PostImages::from_gravatar
-	 * @since 11.3
-	 */
-	public function test_from_gravatar_returns_empty_array_if_post_id_is_null() {
-		$post_id = null;
-		$images  = Jetpack_PostImages::from_gravatar( $post_id );
-
-		$this->assertEmpty( $images );
-	}
-
-	/**
-	 * Test if the array extracted is empty in case post_id is an integer.
+	 * @dataProvider provider_gravatar
 	 *
-	 * @covers Jetpack_PostImages::from_gravatar
-	 * @since 11.3
+	 * @since $$next-version$$
+	 *
+	 * @param int|string|null $post_id  The post ID.
+	 * @param array           $expected The expected array.
 	 */
-	public function test_from_gravatar_returns_empty_array_if_post_id_is_integer() {
-		$post_id = 5;
-		$images  = Jetpack_PostImages::from_gravatar( $post_id );
-
-		$this->assertEmpty( $images );
+	public function test_from_gravatar( $post_id, $expected ) {
+		$image_details = Jetpack_PostImages::from_gravatar( $post_id );
+		$this->assertEquals( $expected, $image_details );
 	}
 
 	/**
-	 * Test if the array extracted has a valid image when sending a valid post.
+	 * Test data for our tests for Jetpack_PostImages::from_gravatar.
 	 *
-	 * @covers Jetpack_PostImages::from_gravatar
-	 * @since 11.3
+	 * @retun array
 	 */
-	public function test_from_gravatar_returns_valid_image() {
-
-		$post_id = $this->factory->post->create();
+	public function provider_gravatar() {
+		$valid_post_id = $this->factory->post->create();
 
-		$images = Jetpack_PostImages::from_gravatar( $post_id );
-
-		$this->assertCount( 1, $images );
-		$this->assertEquals( 'image', $images[0]['type'] );
-		$this->assertEquals( 'gravatar', $images[0]['from'] );
-		$this->assertStringContainsString( 'gravatar.com/avatar', $images[0]['src'] );
-		$this->assertEquals( 96, $images[0]['src_width'] );
-		$this->assertEquals( 96, $images[0]['src_height'] );
-		$this->assertNotEmpty( $images[0]['href'] );
-		$this->assertSame( '', $images[0]['alt_text'] );
+		return array(
+			'invalid (null) post id'                  => array(
+				null,
+				array(),
+			),
+			'post id does not match an existing post' => array(
+				5,
+				array(),
+			),
+			'valid post'                              => array(
+				$valid_post_id,
+				array(
+					array(
+						'type'       => 'image',
+						'from'       => 'gravatar',
+						'src'        => 'http://1.gravatar.com/avatar/?s=96&d=mm&r=g',
+						'href'       => 'http://example.org/?p=4',
+						'src_width'  => 96,
+						'src_height' => 96,
+						'alt_text'   => '',
+					),
+				),
+			),
+		);
 	}
-
 } // end class
```

Down the road, it may be a good task to refactor the whole test file to rely on data providers more. That's for another PR though :)

darssen replied:
Thanks a lot for the suggestion. I pushed a commit following the suggestion

darssen committed:
Improved post images tests by using Data Providers

darssen committed:
Refactored tests for from gravatar.
Data provider was not working as expected for valid post when executed with whole jetpack test suite.

darssen committed:
Merge branch 'trunk' of github.com:Automattic/jetpack into fix/16835-post-images-php-notice-accessing-post-author-of-non-object

darssen commented:
I have updated the tests following the recommendations above regarding the usage of Data Providers.

There was only one caveat. I had to separate the "valid post" test from the Data Provider for two reasons.

1. Due to [documentation](https://phpunit.readthedocs.io/en/8.5/writing-tests-for-phpunit.html) excerpt below, the `$post_id` generated was deleted when executing the whole jetpack tests suite. 
>All data providers are executed before both the call to the setUpBeforeClass() static method and the first call to the setUp() method. Because of that you can’t access any variables you create there from within a data provider. This is required in order for PHPUnit to be able to compute the total number of tests.

2. Since gravatar URL gets a subdomain dynamically ( 0.gravatar.com, 1.gravatar.com, etc ), we cannot compare the whole array with one assert and we need to do individual asserts for each key in the post.

jeherve commented on projects/plugins/jetpack/tests/php/media/test-class.jetpack-post-images.php:
```diff
@@ -633,4 +633,59 @@ public function test_from_story_block_from_post_id_is_correct_array_videopress_w
+	 */
+	public function test_from_gravatar_invalid( $post_id ) {
+		$image_details = Jetpack_PostImages::from_gravatar( $post_id );
+		$this->assertEquals( array(), $image_details );
```
This could be `assertEmpty`, but I don't think that's a blocker! :) Going to approve this now.

jeherve commented:
> One last step: head over to your WordPress.com diff, D86771-code, and deploy it.

As noted in the diff, we'll have to wait until the next release for that. 👍

sdixon194 commented:
We've released a new version of Jetpack on wpcom, however there are still tests failing on that diff - @darssen would you be able to take a look?
